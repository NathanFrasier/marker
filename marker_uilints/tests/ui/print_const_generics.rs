use std::fmt::Debug;

// FIXME: Make sure this cluster duck works
pub struct ArrayPair<T, const N: usize> {
    left: [T; N],
    right: [T; N],
}

impl<T: Debug, const N: usize> ArrayPair<T, N> {
    fn print(&self) {
        println!("left: {:?}", self.left);
        println!("right: {:?}", self.right);
    }
}

fn main() {
    let _ty = ArrayPair {
        left: [1, 2, 3],
        right: [4, 5, 6],
    };
    // let arr = ArrayPair {
    //     left: [1, 2, 3],
    //     right: [4, 5, 6],
    // };
    // arr.print();
}
