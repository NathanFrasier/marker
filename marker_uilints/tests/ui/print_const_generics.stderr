warning: printing item
 --> $DIR/print_const_generics.rs:2:8
  |
2 | struct PrintMeConstGenerics<const N: usize> {
  |        ^^^^^^^^^^^^^^^^^^^^
  |
  = note: Struct(
              StructItem {
                  data: CommonItemData {
                      id: ItemId(..),
                      span: SpanId(..),
                      vis: Visibility {{ /* WIP: See rust-marker/marker#26 */}},
                      ident: Ident {
                          name: "PrintMeConstGenerics",
                          span: Span {
                              source: File(
                                  "$DIR/print_const_generics.rs",
                              ),
                              start: 8,
                              end: 28,
                          },
                      },
                  },
                  generics: SynGenericParams {
                      params: [
                          Const(
                              SynConstParam {
                                  id: GenericId(..),
                                  name: SymbolId(..),
                                  ty: Num(
                                      SynNumTy {
                                          data: CommonSynTyData {
                                              _lifetime: PhantomData<&()>,
                                              span: SpanId(..),
                                          },
                                          numeric_kind: Usize,
                                      },
                                  ),
                                  default: None,
                                  span: SpanId(..),
                              },
                          ),
                      ],
                      clauses: [],
                  },
                  kind: Field(
                      [
                          Field {
                              id: FieldId(..),
                              vis: Visibility {{ /* WIP: See rust-marker/marker#26 */}},
                              ident: SymbolId(..),
                              ty: Array(
                                  SynArrayTy {
                                      data: CommonSynTyData {
                                          _lifetime: PhantomData<&()>,
                                          span: SpanId(..),
                                      },
                                      inner_ty: Num(
                                          SynNumTy {
                                              data: CommonSynTyData {
                                                  _lifetime: PhantomData<&()>,
                                                  span: SpanId(..),
                                              },
                                              numeric_kind: F32,
                                          },
                                      ),
                                      len: Some(
                                          ConstExpr {
                                              expr: Path(
                                                  PathExpr {
                                                      data: CommonExprData {
                                                          _lifetime: PhantomData<&()>,
                                                          id: ExprId(..),
                                                          span: SpanId(..),
                                                      },
                                                      path: AstQPath {
                                                          self_ty: None,
                                                          path_ty: None,
                                                          path: AstPath {
                                                              segments: [
                                                                  AstPathSegment {
                                                                      ident: Ident {
                                                                          name: "N",
                                                                          span: Span {
                                                                              source: File(
                                                                                  "$DIR/print_const_generics.rs",
                                                                              ),
                                                                              start: 63,
                                                                              end: 64,
                                                                          },
                                                                      },
                                                                      generics: SynGenericArgs {
                                                                          args: [],
                                                                      },
                                                                  },
                                                              ],
                                                          },
                                                          target: Generic(
                                                              GenericId(..),
                                                          ),
                                                      },
                                                  },
                                              ),
                                          },
                                      ),
                                  },
                              ),
                              span: SpanId(..),
                          },
                      ],
                  ),
              },
          )
  = note: `#[warn(marker::test_lint)]` on by default

warning: printing item
 --> $DIR/print_const_generics.rs:6:4
  |
6 | fn print_me() -> PrintMeConstGenerics<3> {
  |    ^^^^^^^^
  |
  = note: Fn(
              FnItem {
                  data: CommonItemData {
                      id: ItemId(..),
                      span: SpanId(..),
                      vis: Visibility {{ /* WIP: See rust-marker/marker#26 */}},
                      ident: Ident {
                          name: "print_me",
                          span: Span {
                              source: File(
                                  "$DIR/print_const_generics.rs",
                              ),
                              start: 73,
                              end: 81,
                          },
                      },
                  },
                  generics: SynGenericParams {
                      params: [],
                      clauses: [],
                  },
                  constness: NotConst,
                  syncness: Sync,
                  safety: Safe,
                  is_extern: false,
                  has_self: false,
                  abi: Default,
                  params: [],
                  return_ty: Some(
                      Path(
                          SynPathTy {
                              data: CommonSynTyData {
                                  _lifetime: PhantomData<&()>,
                                  span: SpanId(..),
                              },
                              path: AstQPath {
                                  self_ty: None,
                                  path_ty: None,
                                  path: AstPath {
                                      segments: [
                                          AstPathSegment {
                                              ident: Ident {
                                                  name: "PrintMeConstGenerics",
                                                  span: Span {
                                                      source: File(
                                                          "$DIR/print_const_generics.rs",
                                                      ),
                                                      start: 87,
                                                      end: 107,
                                                  },
                                              },
                                              generics: SynGenericArgs {
                                                  args: [
                                                      Const(
                                                          SynConstArg {
                                                              span: SpanId(..),
                                                              expr: ConstExpr {
                                                                  expr: IntLit(
                                                                      IntLitExpr {
                                                                          data: CommonExprData {
                                                                              _lifetime: PhantomData<&()>,
                                                                              id: ExprId(..),
                                                                              span: SpanId(..),
                                                                          },
                                                                          value: 3,
                                                                          suffix: None,
                                                                      },
                                                                  ),
                                                              },
                                                          },
                                                      ),
                                                  ],
                                              },
                                          },
                                      ],
                                  },
                                  target: Item(
                                      ItemId(..),
                                  ),
                              },
                          },
                      ),
                  ),
                  body_id: Some(
                      BodyId(..),
                  ),
              },
          )

warning: print type test
  --> $DIR/print_const_generics.rs:13:5
   |
13 | /     let _ty: PrintMeConstGenerics<3> = PrintMeConstGenerics {
14 | |         data: [1.0, 1.5, 2.0],
15 | |     };
   | |______^
   |
   = note: Adt(
               SemAdtTy {
                   def_id: TyDefId(..),
                   generics: SemGenericArgs {
                       args: [
                           Const(
                               SemConstArg {
                                   value: ConstValue {{ /* WIP: See rust-marker/marker#179 */}},
                               },
                           ),
                       ],
                   },
               },
           )

warning: 3 warnings emitted

